DESIGN.md

```diff
@@ -161,87 +161,88 @@ const (
 )
 
 const (
-    ContextSignature  = "Signature"
-    ContextSignature1 = "Signature1"
+    AlgorithmPS256   Algorithm = -37 // RSASSA-PSS w/ SHA-256 by RFC 8230
+    AlgorithmPS384   Algorithm = -38 // RSASSA-PSS w/ SHA-384 by RFC 8230
+    AlgorithmPS512   Algorithm = -39 // RSASSA-PSS w/ SHA-512 by RFC 8230
+    AlgorithmES256   Algorithm = -7  // ECDSA w/ SHA-256 by RFC 8152
+    AlgorithmES384   Algorithm = -35 // ECDSA w/ SHA-384 by RFC 8152
+    AlgorithmES512   Algorithm = -36 // ECDSA w/ SHA-512 by RFC 8152
+    AlgorithmEd25519 Algorithm = -8  // PureEdDSA by RFC 8152
 )
 
-var (
-    AlgorithmPS256   *Algorithm
-    AlgorithmPS384   *Algorithm
-    AlgorithmPS512   *Algorithm
-    AlgorithmES256   *Algorithm
-    AlgorithmES384   *Algorithm
-    AlgorithmES512   *Algorithm
-    AlgorithmEd25519 *Algorithm
-)
-
-type Algorithm struct {
-    Name     string
-    Value    int
-    HashFunc crypto.Hash
-}
+type Algorithm int64
+    func (a Algorithm) ComputeHash(data []byte) ([]byte, error)
+    func (a Algorithm) String() string
 
-func I2OSP(x *big.Int, n int) ([]byte, error)
+func I2OSP(x *big.Int, buf []byte) error
 func OS2IP(x []byte) *big.Int
+func RegisterAlgorithm(alg Algorithm, name string, hash crypto.Hash, hashFunc func() hash.Hash) error
+func Sign1(rand io.Reader, signer Signer, protected ProtectedHeader, external, payload []byte) (*Sign1Message, error)
+func Verify1(msg *Sign1Message, verifier Verifier) error
 
 type ProtectedHeader map[interface{}]interface{}
-    func (h *ProtectedHeader) MarshalCBOR() ([]byte, error)
+    func (h ProtectedHeader) Algorithm() (Algorithm, error)

```

Comment:
Could API on line 184 be changed to be more clear to say:
'GetAlgorithm()'

Reply:
@qmuntal proposed `Algorithm()` with a reasonable motivation: https://github.com/veraison/go-cose/pull/17#discussion_r837799695.

Reply:
Sure!